openhands/server/session/session.py

```diff
@@ -124,9 +124,10 @@ async def on_event(self, event: Event):
             return
         if event.source == EventSource.AGENT:
             await self.send(event_to_dict(event))
-        elif event.source == EventSource.USER and isinstance(
-            event, CmdOutputObservation
-        ):
+        #elif event.source == EventSource.USER and isinstance(
+        #    event, CmdOutputObservation
+        #):
+        elif event.source == EventSource.USER:
```

Comment:
Can you please tell, why was this necessary?

Reply:
If there are multiple tabs / users attached to a session, then we need to make sure that each of the tabs gets input from one user as output - otherwise they only see responses rather than the prompts that generated them - so we now echo all events where the source is `USER` rather than just `CmdOutputObservation` instances.

Reply:
We may need to double check that this works as expected in all cases, for example, the lines below:
- exclude ipython obs
- send error obs as AGENT

I know I meant most obs to be ENV source rather than USER, so if they're ENV and nothing unwanted is USER, these should work fine (tm), but it's worth a quick check.

Reply:
I mean, you're saying that this change should only refer to actual "user messages", that is, from a person. The change does all events with source USER though, which may be more than that.

Reply:
Ugh this logic has gotten pretty messy. We're doing a lot of monkeypatch-y things with the events we send down.

IMO the logic should just be `self.send(event_to_dict(event))` in all cases, except maybe for a few little exceptions. (e.g. we had an exception here for user messages before, because those were being handled on the FE)

To Engel's point we should probably make sure that commands sent from the FE don't appear double

Reply:
@tofarr just make sure that user interactions w/ the terminal behave as expected, and we can merge this. Would love to simplify this logic a bit in the future